ssl/statem/extensions.c

```diff
@@ -1731,3 +1747,137 @@ static int final_psk(SSL_CONNECTION *s, unsigned int context, int sent)
+                || !WPACKET_start_sub_packet_u8(pkt))
+            goto err;
+
+        for (i = 0; sc->cert_comp_prefs[i] != TLSEXT_comp_cert_none; i++) {
+            if (!WPACKET_put_bytes_u16(pkt, sc->cert_comp_prefs[i]))
+                goto err;
+        }
+        if (!WPACKET_close(pkt) || !WPACKET_close(pkt))
+            goto err;
+    } else if (!WPACKET_put_bytes_u16(pkt, TLSEXT_TYPE_compress_certificate)
+            || !WPACKET_start_sub_packet_u16(pkt)
+            || !WPACKET_start_sub_packet_u8(pkt)
+# ifndef OPENSSL_NO_BROTLI
+            || !WPACKET_put_bytes_u16(pkt, TLSEXT_comp_cert_brotli)
+# endif
+# if defined(ZLIB)
+            || !WPACKET_put_bytes_u16(pkt, TLSEXT_comp_cert_zlib)
+# endif
+# ifndef OPENSSL_NO_ZSTD
+            || !WPACKET_put_bytes_u16(pkt, TLSEXT_comp_cert_zstd)
+# endif
+            || !WPACKET_close(pkt)
+            || !WPACKET_close(pkt)) {
+        goto err;
+    }
+
+    sc->ext.compress_certificate_sent = 1;
+    return EXT_RETURN_SENT;
+ err:
+    SSLfatal(sc, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+    return EXT_RETURN_FAIL;
+
+#else
+    return EXT_RETURN_NOT_SENT;
+#endif
+}
+
+#if !defined(OPENSSL_NO_COMP) && (defined(ZLIB) || defined(BROTLI) || defined(ZSTD))
+static int tls_comp_idx(SSL_CONNECTION *sc, uint16_t alg)
+{
+    int i;
+
+    for (i = 0; i < TLSEXT_comp_cert_limit; i++)
+        if (sc->cert_comp_prefs[i] == alg)
+            return i;
+    /* if not found, return the limit */
+    return TLSEXT_comp_cert_limit;
+}
+#endif
+
+int tls_parse_compress_certificate(SSL_CONNECTION *sc, PACKET *pkt, unsigned int context,
+                                   X509 *x, size_t chainidx)
+{
+#if !defined(OPENSSL_NO_COMP) && (defined(ZLIB) || !defined(OPENSSL_NO_BROTLI) || !defined(OPENSSL_NO_ZSTD))
+    PACKET supported_comp_algs;
+    unsigned int comp;
+    int idx;
+
+    /* Ignore the extension and don't send compressed certificates */
+    if ((sc->options & SSL_OP_NO_TX_CERTIFICATE_COMPRESSION) != 0)
+        return 1;
+
+    if (!PACKET_as_length_prefixed_1(pkt, &supported_comp_algs)
+            || PACKET_remaining(&supported_comp_algs) == 0) {
+        SSLfatal(sc, SSL_AD_DECODE_ERROR, SSL_R_BAD_EXTENSION);
+        return 0;
+    }
+
+    if (sc->cert_comp_prefs[0] != TLSEXT_comp_cert_none) {
+        /*
+         * The preference array has real values, so take a look at each
+         * value coming in, and compare against our preference list.
+         * Pick the one closest to be beginning (i.e. lowest index)
+         * If a value is not in the list, it gets the limit value
+         * The array is 0 (i.e. "none") terminated
+         * The preference list only contains supported algorithms
+         */
+        while (PACKET_get_1(&supported_comp_algs, &comp)) {
+            if ((idx = tls_comp_idx(sc, (uint16_t)comp)) != TLSEXT_comp_cert_limit)
+                if (idx < tls_comp_idx(sc, sc->ext.compress_certificate_from_peer))
```

Comment:
It kind of feels like we could refactor this to remember the idx from last time and not have to call `tls_comp_idx` on this line every time